pyinteraph/path_analysis.py

```diff
@@ -0,0 +1,473 @@
+    return paths
+
+def get_common_nodes(paths, threshold):
+    """Takes in an list of paths and returns the nodes which are more 
+    common than the provided threshold.
+    """
+    
+    # Get length of longest path
+    maxl = max([len(p) for p in paths])
+    # Convert path list to array where all paths are equal sized
+    # Use the maximum path length as an upper bound
+    paths_array = np.array([p + [0]*(maxl-len(p)) for p in paths])
+    # Get unique nodes
+    unique_nodes = np.unique(paths_array)
+    unique_nodes = unique_nodes[unique_nodes != "0"]
+    # Get node percentage
+    node_count = np.array([(node == paths_array).sum() for node in unique_nodes])
+    node_perc = node_count/node_count.sum()
+    # Select common nodes
+    common_nodes = unique_nodes[node_perc > threshold]
+    common_nodes = list(common_nodes)
+    return common_nodes
+
+def get_graph_from_paths(paths):
+    """Takes in a list of paths and returns a corresponding graph where
+    the weight of each edge is its count.
+    """
+
+    graph = nx.Graph()
+    for path in paths:
+        for i in range(len(path) - 1):
+            # Get edge
+            node1 = path[i]
+            node2 = path[i+1]
+            # Increment weight if edge exists
+            if graph.has_edge(node1, node2):
+                graph[node1][node2]["weight"] += 1
+            # Add edge otherwise
+            else:
+                graph.add_edge(node1, node2, weight = 1)
+    return graph
+
+
+def get_common_edges(graph, threshold):
+    """Takes in a graph where the edge weights are the count of the edge
+    and returns a list of edges which are larger than the provided 
+    threshold.
+    """
+
+    # Initialize arrays of edges and their count
+    edges = np.array(graph.edges)
+    counts = np.array([graph[e[0]][e[1]]["weight"] for e in edges])
+    # Calculate percentage occurance for each path
+    perc = counts/counts.sum()
+    # Choose paths larger than the threshold
+    common_edges = edges[perc > threshold]
+    # Convert array to list of tuples
+    common_edges = list(map(tuple, common_edges))
+    return common_edges
+
+def get_metapath(graph, res_id, res_space, node_threshold, edge_threshold):
+    """Takes in a list of paths, an edge threshold and a node threshold
+    and returns a list of metapaths where each metapath contains all 
+    nodes and edges above their respective thresholds.
+    """
+    
+    # Calculate all shortest paths
+    paths = get_all_shortest_paths(graph, res_id, res_space)
+    # Create graph from path list
+    paths_graph = get_graph_from_paths(paths)
+
+    # Get common nodes
+    common_nodes = get_common_nodes(paths, node_threshold)
+    # Get common edges
+    common_edges = get_common_edges(paths_graph, edge_threshold)
+
+    # Warn if no common nodes found
+    if len(common_nodes) == 0:
+        warn_str = f"No Nodes found with a frequency higher than: {node_threshold}"
+        log.warning(warn_str)
```

Comment:
You can directly return if you do not find either common nodes or common edges, and avoid to do the check on the length of the list afterwards (because the code will never be executed if one of the lists has length 0, since you directly return from the function)